fred2/customdatadlg.cpp

```diff
@@ -0,0 +1,304 @@
+{
+	if (pMsg->message == WM_KEYDOWN) {
+		if (pMsg->wParam == VK_RETURN || pMsg->wParam == VK_ESCAPE) {
+			return TRUE; // Do not process further
+		}
+	}
+	return CDialog::PreTranslateMessage(pMsg);
+}
+
+BEGIN_MESSAGE_MAP(CustomDataDlg, CDialog)
+ON_BN_CLICKED(IDC_CUSTOM_ADD, OnPairAdd)
+ON_BN_CLICKED(IDC_CUSTOM_REMOVE, OnPairRemove)
+ON_BN_CLICKED(IDC_CUSTOM_UPDATE, OnPairUpdate)
+
+ON_BN_CLICKED(ID_OK, OnButtonOk)
+ON_BN_CLICKED(ID_CANCEL, OnButtonCancel)
+ON_WM_CLOSE()
+
+ON_LBN_SELCHANGE(IDC_CUSTOM_DATA_LISTER, OnListerSelectionChange)
+END_MESSAGE_MAP()
+
+/////////////////////////////////////////////////////////////////////////////
+// CustomDataDlg message handlers
+
+BOOL CustomDataDlg::OnInitDialog()
+{
+	CDialog::OnInitDialog();
+
+	// grab the existing list of custom data pairs and duplicate it. We only update it if the user clicks OK.
+	m_custom_data = The_mission.custom_data;
+
+	update_data_lister();
+
+	// Send default name and values into dialog box
+	UpdateData(FALSE);
+
+	return TRUE; // return TRUE unless you set the focus to a control
+				 // EXCEPTION: OCX Property Pages should return FALSE
+}
+
+void CustomDataDlg::OnButtonOk()
+{
+	// now we set the custom data to our copy
+	The_mission.custom_data = m_custom_data;
+
+	CDialog::OnOK();
+}
+
+void CustomDataDlg::OnButtonCancel()
+{
+	if (query_modified()) {
+		int z = MessageBox("Do you want to keep your changes?", "Close", MB_ICONQUESTION | MB_YESNOCANCEL);
+		if (z == IDCANCEL) {
+			return;
+		}
+
+		if (z == IDYES) {
+			OnButtonOk();
+			return;
+		}
+	}
+
+	CDialog::OnCancel();
+}
+
+// this is called when you hit the enter key
+void CustomDataDlg::OnOK()
+{
+	// override MFC default behavior and do nothing
+}
+
+// this is called when you hit the escape key
+void CustomDataDlg::OnCancel()
+{
+	// override MFC default behavior and do nothing
+}
+
+// this is called when you hit the Close button
+void CustomDataDlg::OnClose()
+{
+	OnButtonCancel();
+}
+
+void CustomDataDlg::OnListerSelectionChange()
+{
+	const int index = m_data_lister.GetCurSel();
+
+	if (index == LB_ERR) {
+		// TODO: clear key and data text  edit boxes?
+		return;
+	}
+
+	Assert(m_data_lister.GetCount() > 0);
+	Assert(index >= 0 && index < (int)m_custom_data.size());
+
+	const SCP_string& key = m_lister_keys[index];
+
+	update_text_edit_boxes(key, m_custom_data[key]);
+	update_help_text("No help text provided");
+
+	for (int i = 0; i < (int)Default_custom_data.size(); i++) {
+		if (Default_custom_data[i].key == key) {
+			update_help_text(Default_custom_data[i].description);
+		}
+	}
+
+}
+
+void CustomDataDlg::OnPairAdd()
+{
+	add_pair_entry((int)m_custom_data.size());
+}
+
+void CustomDataDlg::add_pair_entry(const int insert_index)
+{
+	if (!edit_boxes_have_valid_data(false)) {
+		return;
+	}
+
+	CEdit* data_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_DATA);
+	CString data_str;
+	data_edit->GetWindowText(data_str);
+
+	CEdit* key_edit = (CEdit*)GetDlgItem(IDC_CUSTOM_KEY);
```

Comment:
You can leave this one alone.  The hope being that we could eventually just upgrade to qtFRED and not have to deal with this MFC weirdness anymore.